File: 20-Procesamiento de Lenguaje Natural/nlp.py
# -*- coding: utf-8 -*-
"""
Created on Sat Apr  3 03:21:09 2021

@author: axvargas
"""

# Natural language processing
# Import libraries
import pandas as pd
import numpy as np
# Load dataset
dataset = pd.read_csv("Restaurant_Reviews.tsv", sep="\t", quoting=3)

# Limpieza del texto

# =============================================================================
#Esto solo lo ejecutas si nunca has descargado los stopwords... OJO
# import nltk
# nltk.download('stopwords')
# =============================================================================

import re
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()
corpus = [] #Lista donde agregan los datos procesados
for i in range(dataset.shape[0]):
    review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i])#Eliminar todos los caracteres que no sean letras
    review = review.lower() #Convertir a minusculas
    review = review.strip().split()
    #Eliminar stopwords y #Reducir las las palabras a sus infinitivos
    review = [ps.stem(word) for word in review if (word == 'not' or word not in set(stopwords.words('english')))]
    review = " ".join(review)
    corpus.append(review)
    
#Crear el bag of words
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer()#max_features = 1500
X = cv.fit_transform(corpus).toarray()

y = dataset.iloc[:,1].values

#CHOOSE YOUR CLASSIFICATION ALGO

#divide dataset in training and testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=0)

#scaling variables, not neccesary this time

#Crear modelo de clasificacion
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

#PRedicciones de nuestro modelos puesto para region manager-Partner: 6-7
y_pred = classifier.predict(X_test)

#Elaborar matriz de confusion, para contrastar resultados de prediccion
from sklearn.metrics import confusion_matrix
c_matrix = confusion_matrix(y_true = y_test, y_pred = y_pred)
tn, fp, fn, tp = c_matrix.ravel()
precision_ratio = (tn + tp)/(tn+tp+fn+fp)
error_ratio = (fn + fp)/(tn+tp+fn+fp)

File: 20-Procesamiento de Lenguaje Natural/nlp.R
#Natural language processing

#import dataset
dataset_original = read.delim("Restaurant_Reviews.tsv", quote='', stringsAsFactors = FALSE)

# Limpieza del texto
library(tm)
library(SnowballC)
corpus = VCorpus(VectorSource(dataset_original$Review))
corpus = tm_map(corpus, content_transformer(tolower))
# as.character(corpus[[1]]) consultar la primera poscicion del corpus
corpus = tm_map(corpus, removeNumbers)
corpus = tm_map(corpus, removePunctuation)
corpus = tm_map(corpus, removeWords, stopwords(kind='en'))

#Stemming, eliminar dejar palabras raices
corpus = tm_map(corpus, stemDocument)

#Eliminar espacios adicionales
corpus = tm_map(corpus, stripWhitespace)

# Crear el modelo de Bag of words
#dtm : document term matrix
dtm = DocumentTermMatrix(corpus)
dtm = removeSparseTerms(dtm, 0.999)#mantener el 99% de palabras que mas se repiten, el 1% desaparece

#Converitr el dtm a dataframe
dataset = as.data.frame(as.matrix(dtm))
#A;adir la variable independiente al dataset
dataset$Liked = dataset_original$Liked 

#CHOOSE A CLASIOFICATION ALGO

# Encoding the target feature as factor
dataset$Liked = factor(dataset$Liked, levels = c(0, 1))

#divide dataset into training and testing
library(caTools)
set.seed(123)
split = sample.split(dataset$Liked, SplitRatio = 0.80)
training_set = subset(dataset, split==TRUE)
testing_set = subset(dataset, split==FALSE)

library(randomForest)
classifier = randomForest(x = training_set[,-692],
                          y = training_set$Liked,
                          ntree = 50)
y_pred = predict(classifier, newdata = testing_set[,-692])

#Crear la matriz de confusiones
cm = table(testing_set[,692], y_pred)
cm
